compiler/rustc_middle/src/ty/consts/kind.rs

```diff
@@ -42,7 +40,7 @@ pub enum ConstKind<'tcx> {
     Unevaluated(Unevaluated<'tcx>),
 
     /// Used to hold computed value.
-    Value(ConstValue<'tcx>),
+    Value(ty::ValTree<'tcx>),
```

Comment:
:tada: 

Does this mean `ConstValue` can be simplified in a follow-up PR?

Reply:
`ConstantKind` will stay around, as we'll still need to be able to represent some constants that are not an evaluated constant. We'll need to see if just replacing the `Ty` variant with an `Unevaluated` variant suffices.

But yes, feel free to do arbitrary modifications to `ConstValue` after this PR

Reply:
> ConstantKind will stay around, as we'll still need to be able to represent some constants that are not an evaluated constant. We'll need to see if just replacing the Ty variant with an Unevaluated variant suffices.

Didn't you say in the last PR that it was a transitional thing?^^ How does it fit the plan at https://hackmd.io/Qvrj_eOFTkCHZrhJ7f1ItA?

Reply:
well... depends on the level at which we are integrating things. We can add an unevaluated variant to `ConstValue`. I don't have a strong opinion on the way forward, but somewhere in there we'll need an unevaluated variant, and I think the smallest step forward is to replace `Ty` with `Unevaluated`. An independent step would be to eliminate `ConstValue` by merging it into `ConstantKind`.

Reply:
In https://hackmd.io/Qvrj_eOFTkCHZrhJ7f1ItA we had a `LazyConst` (name to be bikeshed) that abstracts the way the "unevaluated" part so it could be reused between the type system and then MIR. Now instead we have `ConstKind`, `ConstValue`, and `ConstantKind`, and I am very confused about how anything fits together.^^

Reply:
In particular, it looks like `mir::ConstantKind` can be a `ty::ConstKind` (via `ty::Const`), so we can have valtrees in the MIR, which seems wrong?

Looks like `mir::ConstantKind` might better be called `mir::Const` since (I think) it corresponds to `ty::Const` but on the MIR level. `ty::ConstKind` exists just because all `ty::Const` have a common field.

Reply:
> In particular, it looks like `mir::ConstantKind` can be a `ty::ConstKind` (via `ty::Const`), so we can have valtrees in the MIR, which seems wrong?

yea, this is all a bit roundabout right now.

* The `ConstantKind::Ty` should go and be replaced by an `Unevaluated` variant (and possibly a `Param` variant)
* We should rename `mir::Constant` to `mir::Const` (and `mir::ConstantKind` to `mir::ConstKind`).
* We should add a `Ty` field to `mir::Const` and remove it from the `mir::ConstKind` varians.

At that point we have pretty much the `LazyConst` design, but without generics and without having variants in the `mir` version that we'll never need (like `Infer`).

Reply:
Okay, we seem to at least roughly agree then. :) `LazyConst` was based on the assumption that there's a bunch of code to share on the `Unevaluated`/`Param`/... side of things; if that's not the case it indeed makes more sense to just specialize them.

> The ConstantKind::Ty should go and be replaced by an Unevaluated variant (and possibly a Param variant)

Once that is done, is the valtree-to-op code even still needed?